contracts/MultiPartyEscrow.sol

```diff
@@ -0,0 +1,272 @@
+    }
+
+    // the recipient can close the channel at any time by presenting a
+    // signed amount from the sender. The recipient will be sent that amount. The recipient can choose: 
+    // send the remainder to the sender (is_sendback == true), or put that amount into the new channel.
+    function channel_claim(uint256 channel_id, uint256 amount, bytes memory signature, bool is_sendback) 
+    public 
+    {
+        PaymentChannel storage channel = channels[channel_id];
+        require(amount <= channel.value);
+        require(msg.sender == channel.recipient);
+ 
+        //message which was signed contains the address of MPE contract ("this"), but we will add it later.
+        require(isValidSignature_claim(channel_id, channel.nonce, amount, signature, channel.sender));
+        
+        balances[msg.sender]       += amount;
+        channels[channel_id].value -= amount;
+    
+        if (is_sendback)    
+            {
+                _channel_sendback_and_reopen(channel_id);
+            }
+            else
+            {
+                //reopen new "channel", without sending back funds to "sender"        
+                channels[channel_id].nonce += 1;
+            }
+    }
+
+
+    /// the sender can extend the expiration at any time
+    function channel_extend(uint256 channel_id, uint256 new_expiration) 
+    public 
+    returns(bool)
+    {
+        PaymentChannel storage channel = channels[channel_id];
+
+        require(msg.sender == channel.sender);
+        require(new_expiration > channel.expiration);
+
+        channels[channel_id].expiration = new_expiration;
+        return true;
+    }
+    
+    /// the sender could add funds to the channel at any time
+    function channel_add_funds(uint256 channel_id, uint256 amount)
+    public
+    returns(bool)
+    {
+        require(balances[msg.sender] >= amount);
+        
+        PaymentChannel storage channel = channels[channel_id];
+        
+        //TODO: we could remove this require and allow everybody to funds it
+        require(msg.sender == channel.sender);
+
+        channels[channel_id].value += amount;
+        balances[msg.sender]       -= amount;
+        return true;
+    }
+
+    function channel_extend_and_add_funds(uint256 channel_id, uint256 new_expiration, uint256 amount)
+    public
+    {
+        require(channel_extend(channel_id, new_expiration));
+        require(channel_add_funds(channel_id, amount));
+    }
+    
+    // sender can claim refund if the timeout is reached 
+    function channel_claim_timeout(uint256 channel_id) 
+    public 
+    {
+        require(msg.sender == channels[channel_id].sender);
+        require(now >= channels[channel_id].expiration);
+        _channel_sendback_and_reopen(channel_id);
+    }
+
+
+    function isValidSignature_open_channel_replaysafe(address recipient, uint256 value, uint256 expiration, uint256 replica_id, uint256 message_nonce, bytes memory signature, address sender)
```

Comment:
I would decouple the logic of isValidSiganture and move the business logic where is used instead of having a specific function. They are both used once in the whole code, useless as specific helpers

Reply:
agree! And it make contract a little bit cheaper :) (35884 vs 36027 ) ... Will commit this evening. In airport now.